sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -137,13 +139,98 @@ public AnalyzedRelation analyze(Node node, Analysis analysis) {
+                selectAnalysis.add(field.path(), field);
+            }
+
+            return new OrderedLimitedRelation(
+                    childRelation,
+                    analyzeOrderBy(
+                        selectAnalysis,
+                        node.getOrderBy(),
+                        expressionAnalyzer,
+                        expressionAnalysisContext,
+                        false,
+                        false),
+                    longSymbolOrNull(node.getLimit(), expressionAnalyzer, expressionAnalysisContext),
+                    longSymbolOrNull(node.getOffset(), expressionAnalyzer, expressionAnalysisContext));
+        }
+        return process(node.getQueryBody(), statementContext);
     }
 
     @Override
     protected AnalyzedRelation visitUnion(Union node, StatementAnalysisContext context) {
-        throw new UnsupportedFeatureException("UNION is not supported");
+        if (node.isDistinct()) {
+            throw new UnsupportedFeatureException("UNION [DISTINCT] is not supported");
+        }
+        QueriedRelation left = (QueriedRelation) process(node.getLeft(), context);
+        QueriedRelation right = (QueriedRelation) process(node.getRight(), context);
+
+        ensureUnionOutputsHaveTheSameSize(left, right);
+        ensureUnionOutputsHaveCompatibleTypes(left, right);
+
+        UnionSelect unionSelect = new UnionSelect(left, right, node.isDistinct());
+        unionSelect.querySpec().outputs(left.querySpec().outputs());
+        return unionSelect;
+    }
+
+    private static void ensureUnionOutputsHaveTheSameSize(QueriedRelation left, QueriedRelation right) {
+        if (left.outputs().size() != right.outputs().size()) {
+            throw new UnsupportedOperationException("Number of output columns must be the same for all parts of a UNION");
+        }
+    }
+
+    private static void ensureUnionOutputsHaveCompatibleTypes(QueriedRelation left, QueriedRelation right) {
+        List<Symbol> leftOutputs = left.outputs();
+        List<Symbol> rightOutputs = right.outputs();
+        for (int i = 0; i < leftOutputs.size(); i++) {
+            Symbol leftSymbol = leftOutputs.get(i);
+            Symbol rightSymbol = rightOutputs.get(i);
+            DataType leftDataType = leftSymbol.valueType();
+            DataType rightDataType = rightSymbol.valueType();
+
+            if (leftDataType.equals(rightDataType)) {
+                continue;
+            } else if (rightDataType.precedes(leftDataType) &&
+                       leftSymbol.valueType().isConvertableTo(rightSymbol.valueType())) {
+                left.outputs().set(i, leftSymbol.cast(rightDataType));
+            } else if (rightSymbol.valueType().isConvertableTo(leftSymbol.valueType())) {
+                rightOutputs.set(i, rightSymbol.cast(leftDataType));
+            } else {
+                throw new UnsupportedOperationException("Corresponding output columns must be compatible for all parts of a UNION");
+            }
```

Comment:
It would be nice not to duplicate logic from the type precedence here. Ideally, we would use `FuncParams` here.